Fix PostEvictionCallbacks bug

adamsitnik opened the pull request:
Explanation: after #45962 the `this` keyword was not referring to an instance of the `CacheEntry` type:

https://github.com/dotnet/runtime/blob/6b4c1ad8b2297bfa0da566f7097ec5ed38fc79a5/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs#L38

And it was causing a silent error after performing an invalid cast as a part of the scheduled task:

https://github.com/dotnet/runtime/blob/dfc70d174ff13894d9ca6added89601394c0ea2a/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L207-L209

Fixes #46774

adamsitnik committed:
add a failing test

adamsitnik committed:
fix the actual bug (this was not a CacheEntry but a CacheEntryTokens instance here)

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/TokenExpirationTests.cs:
```diff
@@ -207,6 +207,33 @@ public void TokenExpiresOnRegister()
+        [Fact]
+        public async Task PostEvictionCallbacksGetInvokedWhenMemoryCacheEntriesExpireWithAnActiveChangeToken()
+        {
+            var cache = new MemoryCache(new MemoryCacheOptions());
```
using?

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/TokenExpirationTests.cs:
```diff
@@ -207,6 +207,33 @@ public void TokenExpiresOnRegister()
+
+            Assert.False(cache.TryGetValue(key, out _));
+
+            void OnEntryEvicted(object key, object value, EvictionReason reason, object state)
```
Nit: use a lambda to move this to be part of the PostEvictionCallbacks line, for clarity?

stephentoub commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/TokenExpirationTests.cs:
```diff
@@ -207,6 +207,33 @@ public void TokenExpiresOnRegister()
+
+            Assert.True(cache.TryGetValue(key, out _));
+
+            await Task.Run(() => tcs.Task.TimeoutAfter(TimeSpan.FromSeconds(10)));
```
Why is the `Task.Run` needed?

adamsitnik committed:
address code review feedback

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/TokenExpirationTests.cs:
```diff
@@ -207,6 +207,29 @@ public void TokenExpiresOnRegister()
+            using var cache = new MemoryCache(new MemoryCacheOptions());
+            var key = new object();
+
+            var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
```
Instead of doing this waiting, can we instead `cts.Cancel()` directly when we want to expire the entry?

eerhardt commented on src/libraries/Microsoft.Extensions.Caching.Memory/tests/TokenExpirationTests.cs:
```diff
@@ -207,6 +207,29 @@ public void TokenExpiresOnRegister()
+
+            Assert.True(cache.TryGetValue(key, out _));
+
+            await tcs.Task.TimeoutAfter(TimeSpan.FromSeconds(10));
```
Do we need a new dependency on TaskTimeoutExtensions? Instead, can we use a `ManualResetEvent` and verify `WaitOne` returns `true`?

eerhardt replied:
Here's what I was thinking:

```C#
            using var cache = new MemoryCache(new MemoryCacheOptions());
            var key = new object();

            var cts = new CancellationTokenSource();
            var callbackInvoked = new ManualResetEvent(false);

            cache.Set(key, new object(), new MemoryCacheEntryOptions
            {
                ExpirationTokens = { new CancellationChangeToken(cts.Token) },
                PostEvictionCallbacks = { new PostEvictionCallbackRegistration {
                    EvictionCallback = (key, value, reason, state) => callbackInvoked.Set() } },
            });

            Assert.True(cache.TryGetValue(key, out _));

            cts.Cancel();

            Assert.True(callbackInvoked.WaitOne(TimeSpan.FromSeconds(10)));
            Assert.False(cache.TryGetValue(key, out _));
```

eerhardt reviewed (approved):
LGTM. Thanks for getting the fix up so quick.

I just had a couple minor test suggestions.

adamsitnik committed:
address code review feedback:

- remove the dependency to TaskTimeoutExtensions
- don't wait 1s, call Cancel in explicit way

pranavkm reviewed (approved):
Thanks for the quick fix!
